File: src/NRustLightning.Server/Startup.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Internal;
using Microsoft.Extensions.Logging;
using NRustLightning.Infrastructure.Configuration;
using NRustLightning.Infrastructure.JsonConverters;
using NRustLightning.Server.Configuration;
using NRustLightning.Server.Interfaces;
using NRustLightning.Server.Middlewares;
#if DEBUG
using Microsoft.OpenApi.Models;
#endif

namespace NRustLightning.Server
{
    public class Startup
    {
        private readonly ILogger<Startup> logger;

        internal static ILoggerFactory GetStartupLoggerFactory()
            => LoggerFactory.Create(builder =>
            {
                builder.AddConsole();
                builder.AddDebug();
                builder.SetMinimumLevel(LogLevel.Debug);
            });
        public Startup(IConfiguration configuration)
        {
            this.logger = GetStartupLoggerFactory().CreateLogger<Startup>();
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddControllers().AddJsonOptions(options =>
            {
                options.JsonSerializerOptions.PropertyNameCaseInsensitive = true;
                options.JsonSerializerOptions.Converters
                    .Add(new HexPubKeyConverter());
                options.JsonSerializerOptions.Converters
                    .Add(new PaymentRequestJsonConverter());
                options.JsonSerializerOptions.Converters.Add(new uint256JsonConverter());
                options.JsonSerializerOptions.Converters.Add(new FeatureBitJsonConverter());
                
                options.JsonSerializerOptions
                    .Converters.Add(new NullableStructConverterFactory());
                options.JsonSerializerOptions.Converters
                    .Add(new JsonFSharpConverter());
            });
            services.AddHttpClient();
            services.AddSingleton<ISystemClock, SystemClock>();
            services.ConfigureNRustLightning(Configuration, logger);
            services.AddNRustLightning();
            services.AddMvc();
            services.ConfigureNRustLightningAuth(Configuration);

#if DEBUG
            services.AddSwaggerGen(c =>
            {
                c.SwaggerDoc("v1", new OpenApiInfo
                {
                    Version = "v1",
                    Title = "NRustLightning API",
                    Description = "API endpoint for NRustLightning.Server",
                    Contact = new OpenApiContact()
                    {
                        Name = "Joe Miyamoto",
                        Email = "joemphilips@gmail.com"
                    },
                    TermsOfService = new Uri("https://example.com/terms"),
                    License = new OpenApiLicense
                    {
                        Name = "Use under MIT license",
                        Url = new Uri("https://opensource.org/licenses/MIT"),
                    }
                });
            });
#endif
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, IHostApplicationLifetime lifetime)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                var useLoggingMiddleware = Configuration.GetSection("debug").GetOrDefault("http", true);
                if (useLoggingMiddleware)
                    app.UseMiddleware<RequestResponseLoggingMiddleware>();

#if DEBUG
                app.UseSwagger();
                app.UseSwaggerUI(c =>
                {
                    c.SwaggerEndpoint("/swagger/v1/swagger.json", "NRustLightning API v1");
                });
#endif
            }
            else
            {
                app.UseHttpsRedirection();
            }

            app.UseRouting();
            
            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
            });
        }
    }
}

File: src/NRustLightning.GUI/src/NRustLightning.GUI.Server/Startup.fs
namespace NRustLightning.GUI.Server

open System.IO
open Microsoft.AspNetCore
open Microsoft.AspNetCore.Authentication.Cookies
open Microsoft.AspNetCore.Builder
open Microsoft.AspNetCore.Hosting
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Bolero.Remoting.Server
open Bolero.Server.RazorHost
open Microsoft.Extensions.Logging
#if DEBUG
open NRustLightning.GUI.Server.HotReloadExtensions
#endif

[<AutoOpen>]
module private Helpers =
    let getStartupLoggerFactory() =
        LoggerFactory.Create(fun builder ->
                builder
                    .AddConsole()
                    .AddDebug()
                    .SetMinimumLevel(LogLevel.Debug) |> ignore
            )

type Startup(config: IConfiguration) =
    let logger = getStartupLoggerFactory().CreateLogger<Startup>();
    // This method gets called by the runtime. Use this method to add services to the container.
    // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
    member this.ConfigureServices(services: IServiceCollection) =
        services.AddMvc().AddRazorRuntimeCompilation() |> ignore
        services.AddServerSideBlazor() |> ignore
        services
            .AddHttpClient()
            .ConfigureNRustLightning(config)
            .AddNRustLightning()
            .AddAuthorization()
            .AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
                .AddCookie()
                .Services
            .AddRemoting<BookService>()
            .AddRemoting<ConfigurationService>()
            .AddRemoting<WalletService>()
            .AddBoleroHost()
#if DEBUG
            .AddHotReload(templateDir = __SOURCE_DIRECTORY__ + "/../NRustLightning.GUI.Client")
#endif
        |> ignore

    // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
    member this.Configure(app: IApplicationBuilder, env: IWebHostEnvironment) =
        app
            .UseAuthentication()
            .UseRemoting()
            .UseStaticFiles()
            .UseRouting()
            .UseBlazorFrameworkFiles()
            .UseEndpoints(fun endpoints ->
#if DEBUG
                endpoints.UseHotReload()
#endif
                endpoints.MapBlazorHub() |> ignore
                endpoints.MapFallbackToPage("/_Host") |> ignore)
        |> ignore

module Program =
    
    let configureConfig (c: IConfigurationBuilder) =
        c
            .SetBasePath(Directory.GetCurrentDirectory())
            .AddJsonFile("appsettings.json", optional=false, reloadOnChange=true)
            .AddJsonFile("appsettings.Development.json", optional=false, reloadOnChange=true) |> ignore

    [<EntryPoint>]
    let main args =
        WebHost
            .CreateDefaultBuilder(args)
            .ConfigureAppConfiguration(configureConfig)
            .UseStaticWebAssets()
            .UseStartup<Startup>()
            .Build()
            .Run()
        0
